Sources/mpc-core-kit-swift/MpcCoreKitSwift.swift

```
@@ -239,43 +216,47 @@ public class MpcCoreKit {
-        }
-        let tssTag = try TssModule.get_tss_tag(threshold_key: tkey!)
-
-        let tssPubKey = try await TssModule.get_tss_pub_key(threshold_key: tkey!, tss_tag: tssTag)
-
-        return MpcKeyDetails(tssPubKey: tssPubKey, metadataPubKey: try finalKeyDetails.pub_key.getPublicKey(format: .EllipticCompress), requiredFactors: finalKeyDetails.required_shares, threshold: finalKeyDetails.threshold, shareDescriptions: finalKeyDetails.share_descriptions, totalShares: finalKeyDetails.total_shares, totalFactors: 0)
+        return try await getKeyDetails()
     }
 
     private func existingUser() async throws {
-        guard tkey != nil else {
+        guard let thresholdKey = tkey else {
             throw CoreKitError.invalidTKey
         }
 
-        // try check for hash factor
-        if option.disableHashFactor == false {
-            let factor = try Utilities.getHashedPrivateKey(postboxKey: oauthKey!, clientID: option.Web3AuthClientId)
-            // TODO: factors need to be verified before insertion
-            try await inputFactor(factorKey: factor)
-            factorKey = factor
-        } else {
-            let factor = try await getDeviceFactor()
-            // TODO: factors need to be verified before insertion
-            try await inputFactor(factorKey: factor)
-            factorKey = factor
+        do {
+            // try check for hash factor
+
+            let hashFactor = try Utilities.getHashedPrivateKey(postboxKey: oAuthKey!, clientID: option.web3AuthClientId)
+            let hashFactorPub = try SecretKey(hex: hashFactor).toPublic().serialize(compressed: true)
+            let allFactorPub = try await getAllFactorPubs()
+
+            if option.disableHashFactor == false && allFactorPub.contains(hashFactorPub) {
+                try await inputFactor(factorKey: hashFactor)
+                factorKey = hashFactor
+            } else {
+                let deviceFactor = try await getDeviceFactor()
```

device factor might not be available. We should throw error in that case

---

Lets check if this function throws if device share is empty as error is being swallowed in catch intentionally

---

getDeviceFactor() does throw with either CoreKitError.metadataPubKeyUnavailable or CoreKitError.notFound if it is not found.